clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -503,6 +503,14 @@ Sender newSender(LogContext logContext, KafkaClient kafkaClient, ProducerMetadat
                 apiVersions);
     }
 
+    private static long metadataMaxIdleMs(ProducerConfig config) {
+        long metadataMaxIdleMs = config.getLong(ProducerConfig.METADATA_MAX_IDLE_CONFIG);
+        if (metadataMaxIdleMs != ProducerConfig.DEFAULT_METADATA_MAX_IDLE) {
```

Comment:
I think the correct behavior should be
1. If there is setting on METADATA_TOPIC_EXPIRY_MS_CONFIG, its value should be returned.
2. If there is no setting on METADATA_TOPIC_EXPIRY_MS_CONFIG, but there is a setting on METADATA_MAX_IDLE_CONFIG, the latter's value should be returned.

Thus I feel it's better to check the LI specific config first, and return its value if not null.

The current implementation would run into a trouble where
METADATA_MAX_IDLE_CONFIG is set to its default value while there is no value set for METADATA_TOPIC_EXPIRY_MS_CONFIG.

Reply:
We have set the default value to be the same for `METADATA_MAX_IDLE_CONFIG` and `METADATA_TOPIC_EXPIRY_MS_CONFIG `. Thus, if `METADATA_MAX_IDLE_CONFIG` is set to a default value `DEFAULT_METADATA_MAX_IDLE`, the code will try to get `METADATA_TOPIC_EXPIRY_MS_CONFIG`. If that is not set, it will fall back to the `DEFAULT_METADATA_MAX_IDLE`. So I don't understand how the current implementation will run into trouble.

Overall, I think that since `METADATA_MAX_IDLE_CONFIG` is a newer configuration being introduced in Kafka 3.0, and we encourage clients to use this rather than the non-standard `METADATA_TOPIC_EXPIRY_MS_CONFIG`, we should try to get that value first, failing which, we fall back to the LI-specific config.

Reply:
Ah, you are right, the trouble case doesn't hold since there are default values on both.

Still, I feel it's easier to understand if there is a defined order of precedence between these two values. Suppose a user finds out about METADATA_MAX_IDLE_CONFIG and isn't aware of the other METADATA_TOPIC_EXPIRY_MS_CONFIG config. And the latter has an value that's quite different from the default, say 60 * 1000.

If the user has this line of code
setProperty(METADATA_MAX_IDLE_CONFIG, 5 * 60 * 1000 + 1)

by changing the value slightly he/she'll get a drastically different result, e.g.
setProperty(METADATA_MAX_IDLE_CONFIG, 5 * 60 * 1000).
I don't quite like the idea of letting the value control the precedence of these two configs.

That being said, if you still think the current implementation is better and can get approval from another dev, I'd be ok with it too.

Reply:
So this is the current behavior -
1. If `METADATA_MAX_IDLE_CONFIG` is defined, use it. If not, progress below.
2. If `METADATA_TOPIC_EXPIRY_MS_CONFIG` is defined, use it. If not, use the default value.

I feel like there is a clearly defined order of precedence here.

We do not expect both values to be defined. We expect
`METADATA_TOPIC_EXPIRY_MS_CONFIG` to be defined for existing clients, and `METADATA_MAX_IDLE_CONFIG` for newer clients. We can, in fact, improve this by throwing an exception if both of these properties have non-default values.

I am not entirely sure of the alternative being suggested. Can you please elaborate?

Reply:
I was thinking of giving a default value of -1 to METADATA_TOPIC_EXPIRY_MS_CONFIG, and then use the following logic
`
long metadataExpiryMs = config.getLong(ProducerConfig.METADATA_TOPIC_EXPIRY_MS_CONFIG);
if (metadataExpiryMs > 0) {
  return metadataExpiryMs;
} else {
  return config.getLong(ProducerConfig.METADATA_MAX_IDLE_CONFIG);
}
`
The caveat with this approach is that if we still want to enforce the atLeast(5000) validation for METADATA_TOPIC_EXPIRY_MS_CONFIG, we'd need to implement a separate validator to check that it's either -1 or >= 5000.
Compared with the original implementation, it treats -1 as the special value instead of 5 * 60 * 1000, which I feel makes the code a little easier to understand. What do you think?

Reply:
I agree with you. Made the requested changes, and added a test to validate all the scenarios (also act as documentation of this discussion).